Merge Variable and Tensor classes

colesbury opened the pull request:
This replaces the torch.Tensor constructors with factories that produce
Variables. Similarly, functions on the torch module (e.g. torch.randn)
now return Variables.

To keep the PR to a reasonable size, I've left most of the unused tensor
code. Subsequent PRs will remove the dead code, clean-up calls to
torch.autograd.Variable, and rename Variable to Tensor everywhere.

There are some breaking changes because Variable and Tensors had
slightly different semantics. There's a list of those changes here:

  https://github.com/pytorch/pytorch/wiki/Breaking-Changes-from-Variable-and-Tensor-merge

colesbury committed:
Merge Variable and Tensor classes

This replaces the torch.Tensor constructors with factories that produce
Variables. Similarly, functions on the torch module (e.g. torch.randn)
now return Variables.

To keep the PR to a reasonable size, I've left most of the unused tensor
code. Subsequent PRs will remove the dead code, clean-up calls to
torch.autograd.Variable, and rename Variable to Tensor everywhere.

There are some breaking changes because Variable and Tensors had
slightly different semantics. There's a list of those changes here:

  https://github.com/pytorch/pytorch/wiki/Breaking-Changes-from-Variable-and-Tensor-merge

colesbury committed:
Fix Python 2 build and test_distributions.py

colesbury committed:
Fix test_cuda.py when CUDA is not available

colesbury committed:
Fix super call for Python 2

colesbury commented:
@pytorchbot retest this please

colesbury commented:
@pytorchbot retest this please

gchanan commented on test/common.py:
```diff
@@ -74,28 +74,30 @@ def wrapper(*args, **kwargs):
+        assert obj.is_leaf
+        t = type_map.get(obj.type(), get_gpu_type(obj.type()))
+        with torch.no_grad():
+            res = obj.clone().type(t)
```
why do you have to clone?  type(t) should create a copy, right?

gchanan commented on test/common_nn.py:
```diff
@@ -263,7 +263,11 @@ def nlllossNd_reference(input, target, weight=None, ignore_index=-100,
-    output = Variable(torch.zeros(out_size).type_as(input))
+    output = torch.zeros(out_size).type_as(input)
+    if isinstance(target, Variable):
+        target = target.data
```
why are these necessary?  I thought .data is basically a no-op.

colesbury replied:
Yeah, it's not necessary here. I was originally trying to separate out some of these changes into separate commits. I'll remove it.

gchanan commented on test/common_nn.py:
```diff
@@ -291,7 +297,8 @@ def nllloss_reference(input, target, weight=None, ignore_index=-100,
     def nll_loss_helper(input, target, weight, ignore_index):
         if target == ignore_index:
-            return (variable(0), variable(0))
+            # TODO: remove once Variable and Tensor are merged
```
aren't they merged now?

gchanan commented on test/common_nn.py:
```diff
@@ -302,12 +309,7 @@ def nll_loss_helper(input, target, weight, ignore_index):
-            losses_tensor = losses_tensor.squeeze(1)
-    else:
-        losses_tensor = torch.Tensor(losses).type_as(input)
+    losses_tensor = torch.Tensor(losses).type_as(input)
```
can we use variable/tensor with dtype here?  (if it's easy for you, it might not be hooked up fully yet, I'd have to check).  I'd just rather avoid introducing more torch.Tensor calls.

gchanan commented on test/common_nn.py:
```diff
@@ -1031,6 +1032,8 @@ def __call__(self, test_case):
             expected_out = self.reference_fn(deepcopy(input),
                                              deepcopy(target), module)
+            if isinstance(expected_out, Variable):
+                expected_out = expected_out.item()
```
why is this necessary?

colesbury replied:
The assertEquals overload we have handles comparisons between tensors differently from comparisons with numbers.

gchanan commented on test/common_nn.py:
```diff
@@ -852,6 +856,8 @@ def __call__(self, test_case):
 
         if self.reference_fn is not None:
             out = test_case._forward(module, input)
+            if isinstance(out, Variable):
```
same question about this .data.

gchanan commented on test/common_nn.py:
```diff
@@ -680,7 +682,7 @@ def _analytical_jacobian(self, module, input, jacobian_input=True, jacobian_para
             if jacobian_input:
                 for jacobian_x, d_x in zip(flat_jacobian_input, iter_tensors(d_input)):
-                    jacobian_x[:, i] = d_x
+                    jacobian_x[:, i] = d_x.contiguous().view(-1)
```
was there a case that broke here?

colesbury replied:
Yes, the assignment didn't obey broadcasting rules.

colesbury replied:
Tensor `__setitem__` didn't fully adhere to broadcasting rules. `Variable.__setitem__` does

gchanan replied:
I'm still missing something -- do we actually need broadcasting here or can we assert that the shapes are the same?

colesbury replied:
We need the `.view()` call. The jacobian is always 2d (`n_inputs x n_outputs`). The gradient d_x can be any shape so we need to view it as 1-d

gchanan commented on test/common.py:
```diff
@@ -221,7 +227,7 @@ def assertTensorsEqual(a, b):
                     diff = a - b
                     diff[nan_mask] = 0
-                    if diff.is_signed():
+                    if diff.is_signed() and 'CharTensor' not in diff.type():
```
we should just implement abs on CharTensor, right?  (Not necessary for this diff).

colesbury replied:
Yes

apaszke replied:
Can we mark this as a TODO/create an issue?

gchanan commented on test/data/cuda_test_ignores.txt:
```diff
@@ -0,0 +1,150 @@
+torch.ByteTensor.abs
```
nit: I'd rename this test_cuda_ignores so it's obvious it only applies to test_cuda and (if it's easy) put a comment to that affect in the file.

colesbury replied:
Good idea. Done

gchanan commented on test/test_nn.py:
```diff
@@ -281,14 +279,27 @@ def __init__(self, *args, **kwargs):
+            def apply_fn(input1, input2, *params):
+                return module(input1, input2, target)
+
+        # We don't pass `target` as part of inputs because we don't currently
```
nit: add a TODO

gchanan reviewed (commented):
looked at most of the tests, many of these questions would probably be answered by reading the rest...

gchanan commented on test/test_torch.py:
```diff
@@ -1391,7 +1398,12 @@ def _test_broadcast(self, cast):
+                # map and map2 are not implementd on CUDA tensors
+                continue
+
+            # TODO: fix masked_scatter and masked_fill broadcasting
```
what exactly is the problem?

colesbury replied:
They use in-place broadcasting rules instead of out-of-place. This is because they're implemented in terms of the out-of-place functions.

colesbury replied:
It's not a breaking change. These didn't exist on Tensor.

gchanan replied:
Good point, I made this a "new issue".

gchanan commented on test/test_torch.py:
```diff
@@ -1391,7 +1398,12 @@ def _test_broadcast(self, cast):
 
-            if hasattr(large_expanded, fn):
+            if small.is_cuda and fn in ['map', 'map2']:
+                # map and map2 are not implementd on CUDA tensors
```
is it easy to throw something in here that will fail if we implement these?  Like a assertRaises or something?  Otherwise people will invent an entirely new test when they implement this instead of just getting rid of this.

colesbury replied:
Adding a new test for these functions seems like a good outcome.

gchanan commented on test/test_torch.py:
```diff
@@ -1718,37 +1648,9 @@ def maybe_squeeze_result(l, r, result):
     def test_broadcast_batched_matmul(self):
         self._test_broadcast_batched_matmul(self, lambda t: t)
 
-    def test_matmul_out(self):
```
why'd we get rid of test_matmul_out?

colesbury replied:
No `matmul` with out argument

gchanan replied:
Ok, added to the list of breaking changes.

gchanan commented on test/test_torch.py:
```diff
@@ -3916,7 +3814,7 @@ def test_index_select(self):
         self.assertEqual(dest.shape, (5, 4, 5))
         for i in range(idx.size(0)):
-            self.assertEqual(dest[i], src[idx[i]])
+            self.assertEqual(dest[i], src[idx[i].item()])
```
is this item necessary?

colesbury replied:
No, not anymore

gchanan commented on test/test_torch.py:
```diff
@@ -4245,7 +4140,7 @@ def test_expand(self):
 
         # test expanding empty to empty
-        self.assertEqual(torch.randn(()).expand(()), torch.randn(()))
+        self.assertEqual(torch.Tensor(0).expand((0,)), torch.Tensor(0))
```
oh wow this is an ugly test.  Is this another incompatibility in that torch.randn(()) used to give an empty tensor but now gives a scalar?  I don't remember this on the incompatibility list.  As before I think we should avoid adding more torch.Tensor calls -- can we do torch.randn(0) here or something similar?

colesbury replied:
I got rid of the torch.Tensor call. I'm not too worried about the change to randn. I don't think anyone is actually likely to use it in a way that will break.

gchanan replied:
Ok, I added the randn thing to the list of incompatibilities under "minor"

gchanan commented on test/test_torch.py:
```diff
@@ -4535,12 +4430,12 @@ def test_nonzero(self):
                     # that the elements flagged positive are indeed non-zero.
                     for i in range(dst1.size(0)):
-                        self.assertNotEqual(tensor[dst1[i, 0], dst1[i, 1]], 0)
+                        self.assertNotEqual(tensor[dst1[i, 0], dst1[i, 1]].item(), 0)
```
this needs a .item()?  (same comment as above)?

colesbury replied:
Yeah, because the way we override assertNotEqual on Tensors to use assertGreatEqual, which calls `abs` which doesn't work on ByteTensors.

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -149,11 +149,20 @@ def should_bind(declaration):
 
 
 def gen_py_torch_functions(out, declarations):
+    def is_namespace_or_type_function(declaration):
```
this name doesn't really match what it's doing (a decl only on the Type and Tensor I would expect to return True from the name); maybe instead of describing what the function does, describe what you are trying to achieve?

gchanan replied:
were you planning to address this?

gchanan commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -89,6 +89,7 @@ static PyMethodDef torch_functions[] = {
   {"clamp", (PyCFunction)THPVariable_clamp, METH_VARARGS | METH_KEYWORDS | METH_STATIC, NULL},
   {"from_numpy", (PyCFunction)THPVariable_from_numpy, METH_STATIC | METH_O, NULL},
   {"variable", (PyCFunction)THPVariable_variable, METH_VARARGS | METH_KEYWORDS | METH_STATIC, NULL},
+  {"spmm", (PyCFunction)THPVariable_mm, METH_VARARGS | METH_KEYWORDS | METH_STATIC, NULL},
```
this isn't used?

colesbury replied:
Huh? `spmm` is used by `test_sparse.py`. We were missing it on Variable

gchanan commented on torch/__init__.py:
```diff
@@ -267,10 +269,9 @@ def storage_type(cls):
-    CharTensor, ByteTensor, HalfTensor
-}
+# The _tensor_classes set is initialized by the call to _C._initialize_tensor_types()
+# which calls torch::tensor::initialize_python_bindings()
```
I think just the first line is sufficient; someone can look at the implementation and see what it does, and this is likely to change and not be updated.

gchanan commented on torch/_tensor_str.py:
```diff
@@ -277,31 +277,39 @@ def _vector_str(self):
-        size_str = 'x'.join(str(size) for size in self.size())
-        size_str_prefix = 'of size ' if self.ndimension() > 0 else 'with no dimension'
+        if empty:
+            size_str = '(0)'
```
Why don't we just use what's in variable.py?
size_str = '(' + ','.join(str(size) for size in self.size()) + (',)' if len(self.size()) == 1 else ')')

This looks like sometimes it's using "x" deliminator, sometimes (), that just seems worse.

colesbury replied:
I got rid of the `x` delimiter

colesbury committed:
Make randperm default to long tensor

colesbury committed:
Clean-ups from review

colesbury committed:
Delete xxx_like tests of out=... arguments

colesbury committed:
Fix zeros_like on sparse tensors

colesbury committed:
HalfTensor fixes in test_cuda.py

colesbury committed:
Create torch.Tensor classes for all defined types

colesbury committed:
Remove assertion in _symbolic_override_wrapper_maker.

The instance_norm passes tensors through kwargs. These are not traced,
but there is no longer a distinction between Tensor and Variable.

colesbury commented on torch/onnx/__init__.py:
```diff
@@ -387,9 +387,7 @@ def wrapper(*args, **kwargs):
-        assert len(list(function._iter_variables_permissive(
-            list(kwargs.values())))) == 0, \
-            "Passing Variable through kwargs is not supported"
+        # TODO: kwargs aren't traced
```
@ezyang what's the correct thing to do here? InstanceNorm passes weight, saved_running_mean, and saved_running_var as kwargs. They used to be Tensors but Tensor and Variable will be the same type.

ezyang replied:
Sigh. Without changing the tracing API, we would need to reconstruct the positional arguments from the keyword arguments, given the argspec of the function, and then pass them on to the tracer. I've written this code once before but it's definitely annoying.

colesbury commented:
@pytorchbot retest this please

colesbury committed:
Use updated mini-sequence labeler example

colesbury committed:
Work around flaky warning handling in Python 2.7

colesbury commented on torch/autograd/gradcheck.py:
```diff
@@ -26,9 +26,11 @@ def zero_gradients(x):
-    if isinstance(input, Variable) and not input.requires_grad:
-        return None
-    elif torch.is_tensor(input) or isinstance(input, Variable):
+    if isinstance(input, Variable):
```
The desired behavior is that if the input should have a gradient we return zeros and None if not.

Previously, it assumed that all Tensors (but not Variables) should have gradients. Now that there is no distinction between the two classes, only Variables that require grad get a jacobian.

gchanan replied:
right, that makes sense.  My question was about the change of behavior for what were Tensors in this test; would we be checking the gradients for those previously but aren't anymore?

colesbury replied:
It doesn't affect anything in test_nn.py or test_autograd.py. I had changed those in previous PRs not to trigger that code path.

colesbury commented on torch/autograd/gradcheck.py:
```diff
@@ -115,7 +116,9 @@ def get_analytical_jacobian(input, output):
                     else:
-                        jacobian_x[:, i] = d_x.to_dense() if d_x.is_sparse else d_x
+                        d_x_dense = d_x.to_dense() if d_x.is_sparse else d_x
+                        assert jacobian_x[:, i].numel() == d_x_dense.numel()
```
I wanted to make sure that it didn't do any non-trivial broadcasting. For example, if `d_x` is only one element it could be broadcastable to almost any shape

gchanan replied:
That gets to my other question; do we need to allow broadcasting at all here?  (i.e. can we just do assert jacobian_x[:, i].shape == d_x_dense.shape?

colesbury replied:
Yes, you can get rid of the assert. It just catches some size mismatch errors a little earlier.

gchanan replied:
That's different than what I mean.  I meant can we make the assert more restrictive, i.e. to check exact shapes rather than number of elements?

gchanan replied:
I'm guessing this is the same as the above -- this worked with tensors because we allowed deprecated broadcasting rules?

gchanan commented on torch/_tensor_str.py:
```diff
@@ -277,31 +277,34 @@ def _vector_str(self):
     if include_footer:
-        size_str = 'x'.join(str(size) for size in self.size())
-        size_str_prefix = 'of size ' if self.ndimension() > 0 else 'with no dimension'
+        size_str = '({})'.format(','.join(str(size) for size in self.shape))
```
Use what's in variable.py:
size_str = '(' + ','.join(str(size) for size in tensor.size()) + (',)' if len(tensor.size()) == 1 else ')')
(this matches numpy behavior in that it prints (1,) instead of (1)

Also does this handle sparse correctly?  It's probably easier to just pull out the include_footer stuff and copy the code from variable.py into here.

gchanan replied:
oh you deleted what was in variable.py; just move it instead.

colesbury replied:
It also prints (,) which is wrong

colesbury replied:
Why don't we just use `str(tuple(self.shape))`?

gchanan replied:
It doesn't print (,).

gchanan replied:
I'm fine with str(tuple(self.shape)); the suggested code doesn't have spaces between elements and it seemed nicer to be more compact but I don't have a strong preference.

gchanan commented on torch/autograd/variable.py:
```diff
@@ -257,18 +245,19 @@ def share_memory_(self):
         and for CUDA tensors. Tensors in shared memory cannot be resized.
         """
         self.storage().share_memory_()
+        return self
```
do we need this?

colesbury replied:
yes

gchanan commented on torch/autograd/variable.py:
```diff
@@ -339,9 +339,10 @@ def __rpow__(self, other):
 
     def __len__(self):
-        return len(self.data)
+        return self.shape[0]
```
this won't work for 0-dim variables.

colesbury replied:
It's not intended to

gchanan commented on torch/csrc/Module.cpp:
```diff
@@ -431,15 +434,21 @@ PyObject *THPModule_addDocStr(PyObject *_unused, PyObject *args)
     if (f->m_ml->ml_doc) {
-      return PyErr_Format(PyExc_RuntimeError,
-          "function '%s' already has a docstring", f->m_ml->ml_name);
+      // temporary workaround
```
add a comment describing what the temporary workaround is and for what.

colesbury replied:
Ooops. This should be removed.

gchanan commented on torch/csrc/Module.cpp:
```diff
@@ -601,6 +614,14 @@ PyObject *THPModule_setFlushDenormal(PyObject *_unused, PyObject *arg) {
   Py_RETURN_TRUE;
 }
 
+static PyObject* THPModule_initializeTensorTypes(PyObject *_unused)
```
why not just call this initializePythonBindings?

colesbury replied:
We have lots of functions that initialize Python bindings of some sort. Open to suggestions for other names, though

gchanan replied:
You are right, I'd be fine with initializeTensorTypeBindings, but if you want to leave it that's okay too.

colesbury committed:
Fix docstrings for torch and torch.Tensor

gchanan commented on torch/functional.py:
```diff
@@ -160,7 +160,7 @@ def btriunpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
         for i in range(nBatch):
             for j in range(sz):
-                k = LU_pivots[i, j] - 1
+                k = int(LU_pivots[i, j] - 1)
```
why not LU_pivots[i, j].item() ?  That would be equivalent to before, right?

colesbury replied:
`LU_pivots[i, j]` is always an integer so `LU_pivots[i, j].item()` and `int(LU_pivots[i, j])` are identical. I've been writing `int(...)` or `float(...)` out of habit because it's backwards compatible with 0.3. This doesn't matter much here, but it's important for the changes to examples.

gchanan commented on torch/legacy/nn/Max.py:
```diff
@@ -23,7 +23,7 @@ def _lazyInit(self):
         if self._indices is None:
             self._indices = \
-                (torch.cuda.LongTensor() if torch.typename(self.output) == 'torch.cuda.FloatTensor'
+                (torch.cuda.LongTensor() if self.output.type() == 'torch.cuda.FloatTensor'
```
nit: dtype (strings suck)

gchanan commented on torch/optim/lbfgs.py:
```diff
@@ -101,7 +101,7 @@ def step(self, closure):
         # evaluate initial f(x) and df/dx
         orig_loss = closure()
-        loss = orig_loss.data[0]
+        loss = float(orig_loss)
```
why not item?  Is the thought that float will be slightly faster?

colesbury replied:
Same speed. I've had a small preference for float()/int() since they work in 0.3 (not particularly important here, but matters for changes to benchmarks/examples).

gchanan commented on torch/csrc/tensor/python_tensor.h:
```diff
@@ -0,0 +1,22 @@
+void set_default_tensor_type(const at::Type& type);
+
+// Same as set_default_tensor_type() but takes a PyObject*
+void py_set_default_tensor_type(PyObject* type_obj);
```
I guess we should have a dtype one of these too?

gchanan commented on torch/csrc/tensor/python_tensor.h:
```diff
@@ -0,0 +1,22 @@
+// etc. and binds them in their containing modules.
+void initialize_python_bindings(PyObject* module);
+
+// Sets the concrete type constructed by calls to torch.Tensor()
```
this also sets the default types for other calls, right (e.g. ones with a dtype parameter)?  Maybe amend the comment to that effect.

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,261 @@
+    throw TypeError("type %s not available", tensor_type.name);
+  }
+  if (tensor_type.is_cuda) {
+    std::call_once(init_cuda_flag, []() {
```
you can use lazy_init.h stuff, right?

colesbury commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -154,8 +153,10 @@ static Tensor legacy_sparse_tensor_ctor(const Type& type, PyObject* args, PyObje
     auto cdata = reinterpret_cast<void*>(r.toInt64(0));
     return type.unsafeTensorFromTH(cdata, true);
   } else if (r.idx == 3) {
+    AutoGPU auto_gpu(r.toInt64(2));
```
It's weird but it matches the previous behavior on Tensors and TestSparse.test_same_gpu requires it.

It seems like the better behavior would be to infer the device form indices/values.

gchanan replied:
I can look into this.  The reason I bring it up is I'm adding dtypes everywhere we have devices, but again, if this is an alias it doesn't make sense.  It's fine to merge it though, this is contained enough.

colesbury committed:
Add error message for len() on 0-d tensor

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,261 @@
+struct PyTensorType {
+  PyTypeObject py_type;
+  at::Type* aten_type;
+  bool is_cuda;
```
it's seems like we should hook this up to the dtype eventually.

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,261 @@
+  at::Type* aten_type;
+  bool is_cuda;
+  bool is_sparse;
+  bool is_default;
```
comment why you have this.

gchanan replied:
also, a name with "all" is probably better than "default".  is_all_types? is_any_type?

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,261 @@
+  HANDLE_TH_ERRORS
+  auto& tensor_type = *((PyTensorType*)type);
+  if (!tensor_type.aten_type) {
+    throw TypeError("type %s not available", tensor_type.name);
```
probably good to special case an error message about cuda;  There's an example in python_arg_parser:
    if (dtype->is_cuda) {
      oss << "  Torch not compiled with CUDA enabled." << std::endl;
    }

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,261 @@
+}
+
+static void py_initialize_tensor_type(PyTypeObject& type, const char* name, PyObject* tp_dict) {
+  // NOTE: we don't use he typical static declaration of PyTypeObject because
```
we don't use he -> we don't use the.

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,261 @@
+static void set_type(PyTensorType& type_obj, Backend backend, ScalarType scalarType) {
+  auto baseType = globalContext().type_registry[static_cast<int>(backend)][static_cast<int>(scalarType)].get();
+  type_obj.aten_type = baseType ? torch::autograd::VariableType::getType(*baseType) : nullptr;
+  type_obj.is_cuda = backend == kCUDA || backend == kSparseCUDA;
```
ya, definitely a lot of duplication with dtypes.

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,261 @@
+
+    PyObject* type_obj = (PyObject*)&tensor_type;
+    Py_INCREF(type_obj);
+    PyModule_AddObject(module_obj.get(), type_name.c_str(), type_obj);
```
check this succeeded?

colesbury committed:
Use Python ints instead of 0-d tensors as DataLoader indices

This removes a data loading slow-down due to the Variable/Tensor merge.
Note that either one of these changes would be sufficient, but they both
seem like good practice to be safe.

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,261 @@
+  }
+  auto type = (PyTensorType*)obj;
+  if (!type->aten_type) {
+    throw TypeError("invalid type object");
```
I think you need an error message as above, with special casing for good cuda error messages.  This can really happen because we create all of the "declared types".

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,274 @@
+    set_name(tensor_type, get_name(backend, scalar_type));
+  }
+
+  set_type(tensor_types.back(), kCPU, kFloat);
```
it's difficult to follow how the original default tensor type is actually initialized (i.e. this just matches up with what we know it is, but doesn't actually set it).  Can we move the initialization in here?

gchanan reviewed (approved):
lgtm once the comments are addressed.

At one point we had discussed (but never came to an agreement) on whether `.data[0]` should translate into a call to `.item()` along with a warning.  Did you consider that further?

apaszke commented on torch/_tensor_docs.py:
```diff
@@ -326,24 +309,20 @@ def add_docstr_all(method, docstr):
 add_docstr_all('copy_',
                r"""
-copy_(src, non_blocking=False, broadcast=True) -> Tensor
+copy_(src, non_blocking=False) -> Tensor
```
Is this included in the breaking change list?

colesbury replied:
I've updated the wiki page: https://github.com/pytorch/pytorch/wiki/Breaking-Changes-from-Variable-and-Tensor-merge

apaszke commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -0,0 +1,274 @@
+  if (THPVariable_Check(arg)) {
+    if (self->is_default) {
+      Py_RETURN_TRUE;
+    }
```
Is default type what we call `torch.Tensor`?

colesbury replied:
yes, exactly. (It's also the default type returned by functions like torch.ones, torch.zeros)

apaszke reviewed (commented):
I skimmed over the implementation and had a couple questions. Can't wait to have this finally merged!

colesbury commented on torch/csrc/autograd/python_variable.h:
```diff
@@ -26,9 +21,18 @@ THP_API PyObject *THPVariableClass;
   return THPVariableClass && PyObject_IsInstance(obj, THPVariableClass);
 }
+
+inline torch::autograd::Variable& THPVariable_Unpack(PyObject* obj) {
```
I was trying to match the Python C-API conventions (i.e. MyType_Check) which we use above.

colesbury commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1299,11 +1299,11 @@ def emit_body(env, option):
             # return the same underlying Tensor type for both real and accreal; this ensures
-            # e.g. x.sum(0) and x.sum() return the same type.
+            # e.g. x.sum(0) and x.sum() return the same type. We explicitly cast to the
+            # ScalarType before constructing the scalarTensor to avoid overflow checking.
```
Mostly because some tests blindly call `x.sum()` on ByteTensors and trigger the check. This isn't great, but it matches the existing behavior of `sum(dim=0)`. Longer term I'd like to:

1) accept an optional dtype when calling sum()
2) probably default to a sufficiently large type (e.g. int64_t for ByteTensor)
3) handle mixed precision operations, at least with scalars, so that things like x / x.sum() don't break

colesbury commented:
Here's the CharTensor issue: https://github.com/pytorch/pytorch/issues/2359. I've also added a TODO in the code.

colesbury committed:
Changes from review:

 * clean-up default tensor type initialization a bit
 * use lazy_init.h
 * use tensor.dtype

colesbury committed:
Rename is_default to is_base_type

colesbury committed:
Improve unavailable type message

colesbury committed:
Fix test_jit.py for scalars

colesbury commented on test/test_jit.py:
```diff
@@ -1409,7 +1409,7 @@ def fn(x, y, z):
         out = []
         for inp in arr:
-            out.append(Variable(torch.from_numpy(np.array([inp], dtype=dtype))))
+            out.append(torch.from_numpy(np.array(inp, dtype=dtype)))
```
cc @jamesr66a @zdevito  

The script seems to process these as scalars (0-d tensors) after Tensor/Variable merge

colesbury committed:
Update is_namespace_or_type_api_function comment

gchanan commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -157,8 +157,12 @@ void FunctionParameter::set_default_str(const std::string& str) {
-      throw std::runtime_error("default value for Type must be none, got: " + str);
+    if (str == "None") {
+      default_type = nullptr;
+    } else if (str == "torch.int64") {
```
I filed https://github.com/pytorch/pytorch/issues/5385 for this.

colesbury committed:
Fix sparse tensor printing

colesbury committed:
Make function name even longer

colesbury committed:
Fix Windows link error

colesbury committed:
Fix Max.py

colesbury commented:
@pytorchbot retest this please

colesbury commented:
I'm going to merge this. The Linux CI has docker issues, but the Windows and Mac OS builds and tests succeed. The linux tests also succeed locally and in previous commits.

There's a perf regression in word_language_model.py (https://github.com/pytorch/pytorch/issues/5388). I think it's mostly in the setup, not in the actual training or evaluation.

dmitriy-serdyuk commented on torch/utils/data/sampler.py:
```diff
@@ -117,7 +117,7 @@ def __init__(self, sampler, batch_size, drop_last):
         batch = []
         for idx in self.sampler:
-            batch.append(idx)
+            batch.append(int(idx))
```
Any particular reason to cast index to integer? It is never documented that the batch sampler accepts only number-like indexes. It was quite handy to pass an object to index a complicated dataset.

colesbury replied:
The issue was that RandomSampler was using `LongTensors` as an index to the dataset. The Variable and Tensor merge had the effect that indexing a 1-d LongTensor return a 0-d (scalar) LongTensor instead of a number. The 0-d tensor was much slower to serialize.

The other part of the fix was this line:
https://github.com/pytorch/pytorch/pull/5225/files/e22441c7d8edd20473033ec8d50e657604b33d02#diff-b8821844d3c39c6b229d352972c72019R50

The change to RandomSampler is probably sufficient. We could probably remove this `int()` call.

dmitriy-serdyuk replied:
Actually, this is addressed in #8652
